crates/burn-tensor/src/tensor/api/int.rs

```diff
@@ -70,6 +71,57 @@ where
         Tensor::new(B::int_into_float(self.primitive))
     }
 
+    /// Produces an indices tensor for the given shape and device.
+    /// The resulting tensor contains coordinates corresponding to each element in the shape at dimension D.
+    ///
+    /// # Arguments
+    ///
+    /// * `shape` - The shape specifying the dimensions of the tensor.
+    /// * `device` - The device to create the tensor on.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `D2` is not equal to `D+1`.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    ///    use burn_tensor::Int;
+    ///    use burn_tensor::{backend::Backend, Shape, Tensor};
+    ///    fn example<B: Backend>() {
+    ///        let device = Default::default();
+    ///        let result = Tensor::<B, 2, Int>::indices::<3>(Shape { dims: [2, 3] }, &device);
+    ///        println!("{}", result);
+    ///    }
+    /// ```
+    pub fn indices<const D2: usize>(shape: Shape<D>, device: &B::Device) -> Tensor<B, D2, Int> {
```

Comment:
I think we should move this method to the backend API so that backends can optimize it, since calling a lot of arange and repeat can be very expansive for big matrices. We should keep default implementation in the backend definition.

Reply:
@McArthur-Alford if you're not sure what that means, see for example the `narrow` op. It is defined by a default implementation but also overridden by some backends.

Reply:
`cartesian_grid` sounds good to me. Ill get started on moving it to a backend op.

Reply:
> @McArthur-Alford if you're not sure what that means, see for example the `narrow` op. It is defined by a default implementation but also overridden by some backends.

Just looking for some clarification on this. It seems like the backend narrow op (i'm specifically looking at `int_narrow`) isn't actually used. Instead the API has its own definition of narrow that doesnt utilize the backend at all? Obviously you can still call `B::int_narrow`,  but my assumption would be that the API should be making calls to the backend implementation, which then themselves have a default implementation?

Reply:
Sorry for the delayed response.

`int_narrow` has a default implementation that works across all backends

```rust
fn int_narrow<const D: usize>(
    tensor: IntTensor<B, D>,
    dim: usize,
    start: usize,
    length: usize,
) -> IntTensor<B, D> {
    narrow::<B, D, Int>(tensor, dim, start, length)
}
```

And for backends that have the op, we override the method (see `int_narrow` in `burn-candle` or `burn-tch` for example).

For this PR, you don't need to override the default implementation for any backend. This can be added later on if required.

Reply:
I was more so wondering about the tensor API narrow function. That also calls  the default narrow, rather than the backend int_narrow/bool_narrow etc. It doesn't seem like the tensor API gets overridden. In that case, overriding the backend op wouldn't actually change what the tensor API is doing right?

Regardless, I should have gotten it working much like narrow, with that last commit.

Reply:
Crap, you're right! I didn't realize the default narrow was called at the tensor API level.. that means it probably never dispatches to the backend even if it's implemented 😅

Reply:
ah, ok. I thought I was going crazy. Glad I helped find a bug at least!